### PR TITLE
fix: inherit OIDC provider email

### PR DESCRIPTION
**Pull request by n1ru4l**
### Background

A user was prompted to confirm their email, but the user never received that email. After a further investigation the users email changed on the OIDC provider side, but we attempted to send the confirmation email to the old users email address.

### Description

This PR solves this edge case by always updating existing records to the latest OIDC provider email upon sign in.
The flow is fully covered with integration tests and as a bonus: some test kit that makes testing oidc flows easier.

### Checklist

- [x] Authentication management
- [x] Testing

### CODE REVIEW

File: packages/services/server/src/supertokens-at-home.ts [hunk 1]
```diff
@@ -1321,7 +1324,7 @@ export async function registerSupertokensAtHome(
           );
           broadcastLog(
             oidcIntegration.id,
-            `an unexpected error occured while calling the user info endoint '${oidcIntegration.userinfoEndpoint}'. HTTP Status: ${grantResponse.status} Body: ${await grantResponse.text()}.`,
+            `an unexpected error occured while calling the user info endoint '${oidcIntegration.userinfoEndpoint}'. HTTP Status: ${grantResponse.status} Body: ${userInfoResponseRaw}.`,
```
Comment (n1ru4l) [inline packages/services/server/src/supertokens-at-home.ts]: this was causing an unexpected exception and was logging the wrong body in the first place 😢

File: packages/services/storage/src/index.ts [hunk 2]
```diff
@@ -758,6 +758,15 @@ export async function createStorage(
...
           }
 
+          if (internalUser.email !== email) {
+            await t.query(sql`
+              UPDATE "users"
+              SET "email" = ${email}
+              WHERE "id" = ${internalUser.id}
+            `);
+            internalUser.email = email;
+          }
```
[the remark this answers is not in the document]
Comment (n1ru4l) [reply packages/services/storage/src/index.ts:768]: no

File: packages/services/server/src/supertokens-at-home.ts [hunk 3]
```diff
@@ -1393,6 +1396,19 @@ export async function registerSupertokensAtHome(
...
+          req.log.debug('providers email has changed. Update record.');
+          user = await supertokensStore.updateOIDCUserEmail({
+            userId: user.userId,
+            newEmail: userInfoBody.data.email,
+          });
+
+          if (!user) {
+            return rep.status(200).send({
+              status: 'SIGN_IN_UP_NOT_ALLOWED',
+              reason: 'Sign in failed. Please contact your origanization administrator.',
```
[the remark this answers is not in the document]
Comment (n1ru4l) [reply packages/services/server/src/supertokens-at-home.ts]: thx bro

File: packages/services/server/src/supertokens-at-home.ts [hunk 4]
```diff
@@ -1321,7 +1324,7 @@ export async function registerSupertokensAtHome(
           );
           broadcastLog(
             oidcIntegration.id,
-            `an unexpected error occured while calling the user info endoint '${oidcIntegration.userinfoEndpoint}'. HTTP Status: ${grantResponse.status} Body: ${await grantResponse.text()}.`,
+            `an unexpected error occured while calling the user info endoint '${oidcIntegration.userinfoEndpoint}'. HTTP Status: ${grantResponse.status} Body: ${userInfoResponseRaw}.`,
           );
```
[the remark this answers is not in the document]
Comment (n1ru4l) [reply packages/services/server/src/supertokens-at-home.ts:1329]: thx

File: packages/services/server/src/supertokens-at-home.ts [hunk 5]
```diff
@@ -1119,7 +1119,7 @@ export async function registerSupertokensAtHome(
...
           return rep.status(200).send({
             status: 'SIGN_IN_UP_NOT_ALLOWED',
-            reason: 'Sign in failed. Please contact your origanization administrator.',
+            reason: 'Sign in failed. Please contact your organization administrator.',
```
Comment (n1ru4l) [inline packages/services/server/src/supertokens-at-home.ts:1122]: sorry for this noise, but AI helped out

File: packages/services/server/src/supertokens-at-home.ts [hunk 6]
```diff
@@ -1393,6 +1396,19 @@ export async function registerSupertokensAtHome(
...
           });
+        } else if (user.email !== userInfoBody.data.email) {
+          req.log.debug('providers email has changed. Update record.');
+          user = await supertokensStore.updateOIDCUserEmail({
+            userId: user.userId,
+            newEmail: userInfoBody.data.email,
+          });
+
+          if (!user) {
+            return rep.status(200).send({
```
Comment (jdolle) [inline packages/services/server/src/supertokens-at-home.ts:1407]: I'd prefer allowing the user to login even if the email update failed.

**Review by jdolle — COMMENTED** [review]
The code looks good--just make sure tests pass

**Comment by n1ru4l** [comment]
@jdolle tests are failing on CI because of differences in accessing the host machine on linux, figuring out right now how to solve it 😢

File: .github/workflows/tests-integration.yaml [hunk 7]
```diff
@@ -55,6 +55,19 @@ jobs:
         with:
           cmd: yq -i 'del(.services.*.volumes)' docker/docker-compose.community.yml
 
+      - name: Show network info
+        run: |
+          ip addr show
+          ip route show
+
```
Comment (n1ru4l) [inline .github/workflows/tests-integration.yaml]:
```suggestion
```

File: docker/docker-compose.end2end.yml [hunk 8]
```diff
@@ -55,5 +55,6 @@ services:
     environment:
       SUPERTOKENS_RATE_LIMIT: '0'
       AUTH_REQUIRE_EMAIL_VERIFICATION: '1'
+
```
Comment (n1ru4l) [inline docker/docker-compose.end2end.yml]:
```suggestion
```

File: .github/workflows/tests-integration.yaml [hunk 9]
```diff
@@ -54,6 +54,13 @@ jobs:
         uses: mikefarah/yq@4839dbbf80445070a31c7a9c1055da527db2d5ee # v4.44.6
         with:
           cmd: yq -i 'del(.services.*.volumes)' docker/docker-compose.community.yml
+      # tests need to access host machine for OIDC stuff
+      - name: patch compose file volumes
```
Comment (n1ru4l) [inline .github/workflows/tests-integration.yaml]:
```suggestion
      - name: make host machine accessible to server container
```
